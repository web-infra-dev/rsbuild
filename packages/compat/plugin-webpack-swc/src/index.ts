export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
