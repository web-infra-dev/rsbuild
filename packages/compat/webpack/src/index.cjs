/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
