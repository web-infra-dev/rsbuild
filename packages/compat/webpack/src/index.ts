export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
