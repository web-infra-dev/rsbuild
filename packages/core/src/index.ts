/**
 * The methods and types exported from this file are considered as
 * the public API of @rsbuild/core.
 */
import { rspack } from '@rspack/core';
import type * as Rspack from '@rspack/core';
import * as __internalHelper from './internal';

// Core methods
export { loadEnv } from './loadEnv';
export { createRsbuild } from './createRsbuild';
export { loadConfig, defineConfig } from './config';

// Rsbuild version
export const version = RSBUILD_VERSION;

// Rspack instance
export { rspack };
export type { Rspack };

// Helpers
export { logger } from '@rsbuild/shared';
export { mergeRsbuildConfig } from './mergeConfig';

// Constants
export { PLUGIN_SWC_NAME, PLUGIN_CSS_NAME } from './constants';

// Types
export type {
  // Config Types
  RsbuildConfig,
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  SourceConfig,
  ServerConfig,
  OutputConfig,
  SecurityConfig,
  PerformanceConfig,
  ModuleFederationConfig,
  // Normalized Config Types
  NormalizedConfig,
  NormalizedDevConfig,
  NormalizedHtmlConfig,
  NormalizedToolsConfig,
  NormalizedSourceConfig,
  NormalizedServerConfig,
  NormalizedOutputConfig,
  NormalizedSecurityConfig,
  NormalizedPerformanceConfig,
  NormalizedModuleFederationConfig,
  // Plugin Types
  RsbuildPlugin,
  RsbuildPlugins,
  RsbuildPluginAPI,
} from './types';

export type {
  RsbuildMode,
  RsbuildEntry,
  RsbuildTarget,
  RsbuildContext,
  RsbuildInstance,
  CreateRsbuildOptions,
  InspectConfigOptions,
  // Subtypes of Config
  Minify,
  Polyfill,
  PrintUrls,
  PublicDir,
  Decorators,
  RspackRule,
  WatchFiles,
  CSSModules,
  CrossOrigin,
  ConsoleType,
  SplitChunks,
  BundlerChain,
  ClientConfig,
  ScriptInject,
  ChainedConfig,
  PostCSSPlugin,
  ScriptLoading,
  LegalComments,
  AliasStrategy,
  FilenameConfig,
  DistPathConfig,
  OutputStructure,
  ChainIdentifier,
  PreconnectOption,
  CSSLoaderOptions,
  ModifyChainUtils,
  StyleLoaderOptions,
  PostCSSLoaderOptions,
  ChainedConfigWithUtils,
  ModifyRspackConfigUtils,
  CSSModulesLocalsConvention,
  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnAfterStartProdServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeStartProdServerFn,
  OnBeforeCreateCompilerFn,
  OnCloseDevServerFn,
  OnDevCompileDoneFn,
  ModifyBundlerChainFn,
  ModifyRspackConfigFn,
  ModifyRsbuildConfigFn,
  TransformFn,
  TransformHandler,
} from '@rsbuild/shared';

export {
  /**
   * @private
   */
  __internalHelper,
};
