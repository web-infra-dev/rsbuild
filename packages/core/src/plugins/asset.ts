import path from 'path';
import {
  getDistPath,
  getFilename,
  chainStaticAssetRule,
} from '@rsbuild/shared';
import type { DefaultRsbuildPlugin } from '@rsbuild/shared';

export function getRegExpForExts(exts: string[]): RegExp {
  const matcher = exts
    .map((ext) => ext.trim())
    .map((ext) => (ext.startsWith('.') ? ext.slice(1) : ext))
    .join('|');

  return new RegExp(
    exts.length === 1 ? `\\.${matcher}$` : `\\.(${matcher})$`,
    'i',
  );
}

export const pluginAsset = (
  assetType: 'image' | 'media' | 'font' | 'svg',
  exts: string[],
): DefaultRsbuildPlugin => ({
  name: `plugin-${assetType}`,

  setup(api) {
    api.modifyBundlerChain((chain, { isProd }) => {
      const config = api.getNormalizedConfig();
      const regExp = getRegExpForExts(exts);
      const distDir = getDistPath(config.output, assetType);
      const filename = getFilename(config.output, assetType, isProd);

      const maxSize = config.output.dataUriLimit[assetType];

      const rule = chain.module.rule(assetType).test(regExp);

      chainStaticAssetRule({
        rule,
        maxSize,
        filename: path.posix.join(distDir, filename),
        assetType,
      });
    });
  },
});
