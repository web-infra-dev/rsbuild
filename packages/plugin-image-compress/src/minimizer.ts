import { Buffer } from 'buffer';
import type { webpack } from '@rsbuild/webpack';
import Codecs from './shared/codecs';
import type { FinalOptions } from './types';

export const IMAGE_MINIMIZER_PLUGIN_NAME =
  '@rsbuild/plugin-image-compress/minimizer' as const;

export interface MinimizedResult {
  source: webpack.sources.RawSource;
}

export class ImageMinimizerPlugin {
  name = IMAGE_MINIMIZER_PLUGIN_NAME;

  options: FinalOptions;

  constructor(options: FinalOptions) {
    this.options = options;
  }

  async optimize(
    compiler: webpack.Compiler,
    compilation: webpack.Compilation,
    assets: Record<string, webpack.sources.Source>,
  ): Promise<void> {
    const cache = compilation.getCache(IMAGE_MINIMIZER_PLUGIN_NAME);
    const { RawSource } = compiler.webpack.sources;
    const { matchObject } = compiler.webpack.ModuleFilenameHelpers;

    const buildError = (error: unknown, file?: string, context?: string) => {
      const cause = error instanceof Error ? error : new Error();
      const message =
        file && context
          ? `"${file}" in "${context}" from Image Minimizer:\n${cause.message}`
          : cause.message;
      const ret = new compiler.webpack.WebpackError(message);

      if (error instanceof Error) {
        (ret as any).error = error;
      }

      return ret;
    };

    const codec = Codecs[this.options.use];
    if (!codec) {
      compilation.errors.push(
        buildError(new Error(`Codec ${this.options.use} is not supported`)),
      );
    }
    const opts = { ...codec.defaultOptions, ...this.options };

    const handleAsset = async (name: string) => {
      const info = compilation.getAsset(name)?.info;
      const fileName = name.split('?')[0];

      // 1. Skip double minimize assets from child compilation
      // 2. Test file by options (e.g. test, include, exclude)
      if (info?.minimized || !matchObject(opts, fileName)) {
        return;
      }

      const { source: inputSource } = compilation.getAsset(name)!;

      const eTag = cache.getLazyHashedEtag(inputSource);
      const cacheItem = cache.getItemCache(name, eTag);
      let result = await cacheItem.getPromise<MinimizedResult | undefined>();

      try {
        if (!result) {
          const input = inputSource.source();
          const buf = await codec.handler(Buffer.from(input), opts);
          result = { source: new RawSource(buf) };
          await cacheItem.storePromise(result);
        }
        compilation.updateAsset(name, result.source, { minimized: true });
      } catch (error) {
        compilation.errors.push(buildError(error, name, compiler.context));
      }
    };
    const promises = Object.keys(assets).map((name) => handleAsset(name));
    await Promise.all(promises);
  }

  apply(compiler: webpack.Compiler) {
    const handleCompilation = (compilation: webpack.Compilation) => {
      compilation.hooks.processAssets.tapPromise(
        {
          name: this.name,
          stage:
            compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
          additionalAssets: true,
        },
        (assets) => this.optimize(compiler, compilation, assets),
      );

      compilation.hooks.statsPrinter.tap(this.name, (stats) => {
        stats.hooks.print
          .for('asset.info.minimized')
          .tap(
            '@rsbuild/plugin-image-compress',
            (minimized, { green, formatFlag }) =>
              minimized && green && formatFlag
                ? green(formatFlag('minimized'))
                : '',
          );
      });
    };
    compiler.hooks.compilation.tap(this.name, handleCompilation);
  }
}
