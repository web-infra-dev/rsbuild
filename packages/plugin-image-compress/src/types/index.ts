import type { Buffer } from 'buffer';
import type {
  JpegCompressOptions,
  PNGLosslessOptions,
  PngQuantOptions,
} from '@napi-rs/image';
import type { Config as SvgoConfig } from 'svgo';

export type ArrayOrNot<T> = T | T[];

export interface WebpTransformOptions {
  quality?: number;
}

export interface CodecBaseOptions {
  jpeg: JpegCompressOptions;
  png: PngQuantOptions;
  pngLossless: PNGLosslessOptions;
  ico: Record<string, never>;
  svg: SvgoConfig;
}

export interface BaseCompressOptions<T extends Codecs> {
  use: T;
  test?: ArrayOrNot<RegExp>;
  include?: ArrayOrNot<RegExp>;
  exclude?: ArrayOrNot<RegExp>;
}

export type FinalOptionCollection = {
  [K in Codecs]: BaseCompressOptions<K> & CodecBaseOptions[K];
};

export type FinalOptions = FinalOptionCollection[Codecs];

export interface Codec<T extends Codecs> {
  handler: (buf: Buffer, options: CodecBaseOptions[T]) => Promise<Buffer>;
  defaultOptions: Omit<FinalOptionCollection[T], 'use'>;
}

export type Codecs = keyof CodecBaseOptions;

export type OptionCollection = {
  [K in Codecs]: K | FinalOptionCollection[K];
};

export type Options = OptionCollection[Codecs];
