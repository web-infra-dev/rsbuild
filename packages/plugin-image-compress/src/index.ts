import type { RsbuildPlugin } from '@rsbuild/core';
import assert from 'node:assert';
import { ImageMinimizerPlugin } from './minimizer';
import { withDefaultOptions } from './shared/utils';
import type { Codecs, Options } from './types';

export type PluginImageCompressOptions = Options[];
export const DEFAULT_OPTIONS: Codecs[] = ['jpeg', 'png', 'ico'];

export interface IPluginImageCompress {
  (...options: Options[]): RsbuildPlugin;
  (options: Options[]): RsbuildPlugin;
}

const castOptions = (args: (Options | Options[])[]): Options[] => {
  const head = args[0];
  // expect [['png', { use: 'jpeg' }]]
  if (Array.isArray(head)) {
    return head;
  }
  // expect ['png', { use: 'jpeg' }]
  const ret: Options[] = [];
  for (const arg of args) {
    assert(!Array.isArray(arg));
    ret.push(arg);
  }
  return ret;
};

const normalizeOptions = (options: Options[]) => {
  const opts = options.length ? options : DEFAULT_OPTIONS;
  const normalized = opts.map((opt) => withDefaultOptions(opt));
  return normalized;
};

/** Options enable by default: {@link DEFAULT_OPTIONS} */
export const pluginImageCompress: IPluginImageCompress = (
  ...args
): RsbuildPlugin => ({
  name: 'rsbuild:image-compress',

  setup(api) {
    const opts = normalizeOptions(castOptions(args));

    api.modifyBundlerChain((chain, { isProd }) => {
      if (!isProd) {
        return;
      }

      chain.optimization.minimize(true);

      for (const opt of opts) {
        chain.optimization
          .minimizer(`image-compress-${opt.use}`)
          .use(ImageMinimizerPlugin, [opt]);
      }
    });
  },
});
