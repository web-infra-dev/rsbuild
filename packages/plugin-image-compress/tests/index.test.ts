import { createStubRsbuild } from '@rsbuild/test-helper';
import { pluginAsset } from '@rsbuild/core/plugins/asset';
import { pluginImageCompress } from '../src';

process.env.NODE_ENV = 'production';

describe('plugin-image-compress', () => {
  it('should generate correct options', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
      plugins: [pluginAsset(), pluginImageCompress()],
    });
    expect(await rsbuild.unwrapConfig()).toMatchSnapshot();
  });

  it('should accept `...options: Options[]` as parameter', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
      plugins: [pluginAsset(), pluginImageCompress('jpeg', { use: 'png' })],
    });
    const config = await rsbuild.unwrapConfig();
    expect(config.optimization?.minimizer).toMatchInlineSnapshot(`
      [
        ModernJsImageMinimizerPlugin {
          "name": "@rsbuild/plugin-image-compress/minimizer",
          "options": {
            "test": /\\\\\\.\\(jpg\\|jpeg\\)\\$/,
            "use": "jpeg",
          },
        },
        ModernJsImageMinimizerPlugin {
          "name": "@rsbuild/plugin-image-compress/minimizer",
          "options": {
            "test": /\\\\\\.png\\$/,
            "use": "png",
          },
        },
      ]
    `);
  });

  it('should accept `options: Options[]` as parameter', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
      plugins: [pluginAsset(), pluginImageCompress(['jpeg', { use: 'png' }])],
    });
    const config = await rsbuild.unwrapConfig();
    expect(config.optimization?.minimizer).toMatchInlineSnapshot(`
      [
        ModernJsImageMinimizerPlugin {
          "name": "@rsbuild/plugin-image-compress/minimizer",
          "options": {
            "test": /\\\\\\.\\(jpg\\|jpeg\\)\\$/,
            "use": "jpeg",
          },
        },
        ModernJsImageMinimizerPlugin {
          "name": "@rsbuild/plugin-image-compress/minimizer",
          "options": {
            "test": /\\\\\\.png\\$/,
            "use": "png",
          },
        },
      ]
    `);
  });
});
