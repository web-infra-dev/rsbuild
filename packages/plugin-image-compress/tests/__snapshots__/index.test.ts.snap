// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-image-compress > should generate correct options 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/image/[name].[contenthash:8][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/image/[name].[contenthash:8][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:png\\|jpg\\|jpeg\\|gif\\|bmp\\|webp\\|ico\\|apng\\|avif\\|tiff\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/svg/[name].[contenthash:8].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/svg/[name].[contenthash:8].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/media/[name].[contenthash:8][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/media/[name].[contenthash:8][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:mp4\\|webm\\|ogg\\|mp3\\|wav\\|flac\\|aac\\|mov\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/font/[name].[contenthash:8][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/font/[name].[contenthash:8][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 10000,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:woff\\|woff2\\|eot\\|ttf\\|otf\\|ttc\\)\\$/i,
      },
    ],
  },
  "optimization": {
    "minimize": true,
    "minimizer": [
      ImageMinimizerPlugin {
        "name": "@rsbuild/plugin-image-compress/minimizer",
        "options": {
          "test": /\\\\\\.\\(\\?:jpg\\|jpeg\\)\\$/,
          "use": "jpeg",
        },
      },
      ImageMinimizerPlugin {
        "name": "@rsbuild/plugin-image-compress/minimizer",
        "options": {
          "test": /\\\\\\.png\\$/,
          "use": "png",
        },
      },
      ImageMinimizerPlugin {
        "name": "@rsbuild/plugin-image-compress/minimizer",
        "options": {
          "test": /\\\\\\.\\(\\?:ico\\|icon\\)\\$/,
          "use": "ico",
        },
      },
    ],
  },
}
`;
